santoku/utils/README.md

```diff
@@ -0,0 +1,111 @@
+nested_boolean_setting = configuration_manager.get_setting("object_setting")\
+                                              .get_setting("nested_boolean_setting")
+
+# alternatively, you can simply pass the list of keys in order
+boolean_setting = configuration_manager.get_setting(
+    ["object_setting", "nested_boolean_setting"]
+)
+```
+
+
+### Schema validation
+For schema validation of configuration files, we use our own definition of what a configuration looks like and also our own definition of what a schema should look like. In short, our schema is a narrow subset of [JSON Schema](https://json-schema.org/). We plan on fully adopting JSON Schema for our validation needs, but due to the lack of native Python tools and the effort of implementing our own parser we erred on the side of something simpler that allows for faster iteration.
+
+```JSON
+# config.json
+[
+    {
+        "name": "configuration_1",
+        "settings": {
+            "boolean_setting": true,
+            "numeric_setting": 42,
+            "object_setting": {
+                "nested_text_setting": "value",
+                ...
+            }
+            ...
+        }
+    },
+    {
+        "name": "configuration_2",
+        "settings": ...
+    },
+    ...
+]
+```
+
+The settings in each configuration must take the form of a JSON object without the use of any array, i.e. a (possibly nested) dictionary. We define the following schema, which is (almost, it uses Python type names) a strict subset of JSON Schema:
+
+```JSON
+# schema.json
+{
+    "boolean_setting": {"type": "bool"},
+    "int_setting": {"type": "int"},
+    "float_setting": {"type": "float"},
+    "string_setting": {"type": "string"},
+    "object_setting": {
+        "nested_boolean_setting": {"type": "bool"},
+        ...
+    },
+    ...
+}
+```
+
+### Recursive type hints
+This module would greatly benefit from recursive type hints. We tried but discovered that currently mypy does not support this. To summarize, it is possible to define custom types using type aliasing, like so:
+
+```python
+from typing import List, Union
+
+# a list of str and/or int
+MyCustomType = List[Union[str, int]]
+```
+
+Recursive types can be created with the restriction that, to use a type inside its own definition, one must use a string literal instead of simply the type. This is called a [forward reference](https://www.python.org/dev/peps/pep-0484/#forward-references).
+
+```python
+from typing import Dict, List, Union
+
+# use "JSON" instead of just JSON
+JSON = Union[List["JSON"], Dict[str, Union[bool, int, float, str, "JSON"]]
+```
+
+Since this issue is currently [the most demanded from the mypy project](https://github.com/python/mypy/issues/731), it is expected that we will eventually have support for recursive type checking. However, even if mypy raises and error, the code above does compile. For now, we will have to make do with the less powerful (i.e. wrong) version:
```

Comment:
Why is highlighting this paragraph as if it was code?

Reply:
Does not happen in VSCode so it is a GitHub thing. It might be that the three dots in the JSON-style code block get it confused?? Idk...